writer_test.go

```diff
@@ -11,16 +24,113 @@ import (
 func TestRecordMetadata(t *testing.T) {
 	is := is.New(t)
 
-	sdkMetadata := sdk.Metadata{
+	var rec sdk.Record
+	rec.Metadata = sdk.Metadata{
 		"created_at":     "2023-03-15T14:25:07Z",
 		"pinecone.prop1": "a1",
 		"pinecone.prop2": "a2",
 	}
-	recMetadata, err := recordMetadata(sdkMetadata)
+	recMetadata, err := parsePineconeMetadata(rec)
 	is.NoErr(err)
 
-	fmt.Println()
-
 	is.Equal(recMetadata.Fields["prop1"].AsInterface().(string), "a1")
 	is.Equal(recMetadata.Fields["prop2"].AsInterface().(string), "a2")
 }
+
+func testRecords(op sdk.Operation, keys ...string) []sdk.Record {
+	recs := make([]sdk.Record, len(keys))
+	for i := range recs {
+		var position sdk.Position
+		var key sdk.Data = sdk.RawData(keys[i])
+		var metadata sdk.Metadata
+		var payload sdk.Data = sdk.StructuredData{
+			"vector": []float64{1, 2},
+		}
+
+		rec := sdk.Record{
+			Position: position, Operation: op,
+			Metadata: metadata, Key: key,
+			Payload: sdk.Change{
+				Before: nil,
+				After:  payload,
+			},
+		}
+		recs[i] = rec
+	}
+
+	return recs
+}
+
+func assertUpsertBatch(is *is.I, batch recordBatch, keys ...string) {
+	upsertBatch, ok := batch.(upsertBatch)
+	is.True(ok) // batch isn't upsertBatch
+
+	is.Equal(len(upsertBatch.vectors), len(keys))
+
+	for i, vec := range upsertBatch.vectors {
+		is.Equal(vec.Id, keys[i])
+	}
+}
```

Comment:
Thank you very much for these changes! I'm sorry if I'm too nitpicky, but I'd suggest checking the vectors too. Having the IDs checked is great and helps a lot, but we should also be sure that the vectors are good. One more reason why I'm suggesting this is that this unit tests helps us check more cases then the integration test (and much quicker).

Reply:
Don't worry I'm getting used to this :D Technically I should be the one apologising, I'm kind of using you as a linter. I agree with most of these nitpicks.

In this case testing the actual contents of the batch itself makes for a much bigger test. Another way to test the correct vector contents are written would be to continue testing that the vector ids are written into the batches, and then test that `parsePineconeVector` does as expected.

I believe this way you test both things, because with the current batch test I make sure that the `parsePineconeVector` is called to get the ids from the vecs, and the `parsePineconeVector` test would make sure that the vector is properly parsed.

Also, there's an integration at `destination_test.go` that asserts that the vec contents are written correctly.

What do you think?

Reply:
Commit #418a78a adds the test